api/server/router/container/backend.go

```diff
@@ -43,7 +43,7 @@ type stateBackend interface {
 	ContainerStop(name string, seconds int) error
 	ContainerUnpause(name string) error
 	ContainerUpdate(name string, hostConfig *container.HostConfig) ([]string, error)
-	ContainerWait(name string, timeout time.Duration) (int, error)
+	ContainerWait(name string, timeout time.Duration, waitForRemoval bool) (int, error)
```

Comment:
> For the other matter, during detach, we should just block, the same way we do today, until we get the die event. 

Do you mean that if client detach, it should wait container until it's dead and removed? I think we need to change this behavior. Moving `--rm` to daemon will grant daemon the capability to do garbage collection. If client detach, it can detach and let daemon handle the left, it doesn't need to block any more.

I just realized @thaJeztah mentioned the same issue: https://github.com/docker/docker/pull/22898#issuecomment-220990054 , it's really confusing and easily treated as a bug.